### ISSUE TITLE
vic-machine reconfigure: Update or reset the base image size

### ISSUE THREAD

**Issue by chengwang86** [vmware/vic#4520]
**User Statement:**

As a VIC admin, I need the ability to update or reset the base image size from which all other images are created. The default value is 8GB. See relevant discussions here #1781 and #159 .

**Details:**
- supported operations:
  - update base image size
- hot update:
  - no; need to restart the VCH to make the changes take effect.
- input/output data format:
  - input: e.g., `--modify --base-image-size ...`
  - output: N/A
- impact on VCH/containerVMs:
  - VCH: This operation needs support from portlayer storage. 
  - containerVMs: no
- is upgrade needed: 
   -  no, since no VCH configuration changes. 

**Acceptance Criteria:**
-  add the ability to update or reset the base image size; the changes will be applied after VCH restart
- add relevant test cases

**Comment by emlin** [comment]
@fdawg4l this base image size is used to set scratch vmdk size only, can you assume in which case, user need to update this size?

**Comment by emlin** [comment]
@chengwang86 portlayer does not resize scratch vmdk file after it's created at this time.

**Comment by fdawg4l** [comment]
Resizing scratch may not be possible at all.  We have several open issues we'd need to address to do this (and we can capture each in an epic since it wouldn't be a trivial change if it is possible at all).

Some questions I have (@hickeng may or may not already have the answers to):
1. Does this `reconfigure` operation need to shrink the base image ever?
    a.  What is this expected impact when a layer (image) exists in the image store which is larger than the expected size to shrink to?  Same for disks attached to containers.
1. Does expanding the base disk require fixing up every image in the image store to reflect the new size?  Since our images are trees which are themselves deltas applied to the scratch base disk, changes in disk geometry (of the base scratch image) have a trickle affect on the leaves (images).  We need to scope what that fixup (if at all) is.  The same fixup operation needs to be answered for container images.
1.  Is the `reconfigure` operation expected to be run on a VCH with containers on-line or offline?
    a.  If online, changing the disk geometry of a delta disk won't be possible since the base disk (scratch) is locked.


Obviously the change which requires the least amount of effort (i.e. no fixup) is preferable.  We may need to migrate from `ext` filesystems to something like `btrfs` or `xfs` which don't require strict volume to filesystem mappings.  If we can answer which fs allows online resize AND plays nice with delta disks (without requiring a fixup), then that's the most preferable way to implement this feature.

**Comment by emlin** [comment]
thanks @fdawg4l for the detailed message. What I'm thinking is if we can change the scratch vmdk file size, without any impact on delta image. 
Just curious, do all vmdk files have same disk size? And is that why we set the scratch disk size?

anyway, for my preference, we need to avoid operations in reconfigure to affect any containers, no matter it's online or offline.

**Comment by fdawg4l** [comment]
> anyway, for my preference, we need to avoid operations in reconfigure to affect any containers, no matter it's online or offline.

Yeah, agreed.  This makes sense.   I'm not sure it's possible though.  :-)

> Just curious, do all vmdk files have same disk size? And is that why we set the scratch disk size?

All images are "deltas" applied to `scratch`.  When you `pull` an image, we find the parent of the image from docker hub, then find its parent and repeat until we find `scratch` in the image's manifest.  Then, we create a delta disk from scratch, write the first image in the chain, close the delta disk, then create a delta disk _from the previous delta disk_ and apply the next image.  We do this until we get down to the last child and return this to the user.  You can now create containers from any image in the tree by creating _another_ delta disk from the image delta disk, and attaching it to the container.

So, the result is all images are related to `scratch`.  `scratch` is the root of the tree and its size is the size we set for all images which get pulled.   We _can_ change the size of the delta when it's being pulled because we're creating a new delta disk at that time.  We would attach the new image's new delta disk with a new size, fix the filesystem size since the block device size is now different, then extract the new image.  So growing should not be a problem.  Shrinking is also possible.  But every disk which is created from the new sized delta disk will inherit that size.

The question then is whether reconfiguring the scratch size is possible and what the effect is on the images since they all derive from it.  And then we need to find out what offline operations need to be done on the images if the size of scratch does change.

But, I think the above may be going about the problem and the fix incorrectly.  What I think we should be doing is allow a new container's disk to be configured when the container is created.  The user probably wants to be able to specify what size disk the container's disk is.  The images don't take up much space since they are "thin" anyway.  That means they only use as much a space as the amount of data which gets written to it.  We can't optimize the size beyond that;  it's as good as it gets.  If we allow the user the option to set the size of the disk the container is created with, then we can create a delta disk for the container (like we already do) but pass the size the user requests for it, attach the disk to the container, and then on container start, we can fix the filesystem size to allow the user to make use of all of the space requested if they want it.

**Comment by emlin** [comment]
thanks @fdawg4l for this explanation, this makes a lot sense to me.

**Comment by emlin** [comment]
so the suggestion is not to change scratch disk size

**Comment by fdawg4l** [comment]
My preference is we leave the the base image (`scratch`) alone unless a user requirement forces us to implement the feature.

I think above all, we need to clarify what the user wants.  Then we can find a way to implement the feature without breaking the disk chains (which is what we really want to avoid).

So, at first pass, we need to 
 1.  Identify what the user wants.
     1. Does s/he want to twiddle the base image size to save logical space?  Or is twiddling the size a response to some maintanance operation (e.g. switching from large rotating disks to smaller SSDs, necessitating the small logical image sizes)?
     1.  Does s/he really just need more space in the container?
 1. Identify what's possible in vsphere from (1).
     1.  Can we change the base image size without effecting the delta disks?  If we have to twiddle the delta disks, then we'll need to twiddle the container read write disks which could be bad.  We'll need to scope this.
     1.  Can we start with a very small base disk, then use a larger logical sized delta for each image?  This is probably the ideal scenario if it's possible.
     1. Scope the effects on VSAN, NFS backed datastores, ISCSI backed datastores, etc.  We found that modifying disks using the FileManager in certain deployment scenarios would convert the thin images to thick disks.  It's likely there are other paths in the vsphere API which do similar things under the covers and we just need to identify if we'll hit these issues.

**Comment by emlin** [comment]
I like the idea of small base disk, and larger delta disk for each image. And right, we need more investigation for vSphere behavior as well.
